Initial implementation of ForwardCRD plugin (#4512)
miekg/coredns: plugin/forwardcrd/README.md

# forwardcrd

## Name

*forwardcrd* - enables proxying DNS messages to upstream resolvers by reading

> CRDs from a Kubernetes cluster

Reviewer: ```suggestion
the `DNSZone` CRD from a Kubernetes cluster
```

matches or includes the zone as a suffix. If **ZONES** are specified it allows
any zone listed as a suffix.

```
forwardcrd [ZONES...] {
  endpoint URL
  tls CERT KEY CACERT
  kubeconfig KUBECONFIG [CONTEXT]
  namespace NAMESPACE
}
```

* `endpoint` specifies the **URL** for a remote k8s API endpoint.  If omitted,
  it will connect to k8s in-cluster using the cluster service account.
* `tls` **CERT** **KEY** **CACERT** are the TLS cert, key and the CA cert file
  names for remote k8s connection.  This option is ignored if connecting
  in-cluster (i.e. endpoint is not specified).
* `kubeconfig` **KUBECONFIG [CONTEXT]** authenticates the connection to a remote
  k8s cluster using a kubeconfig file.  **[CONTEXT]** is optional, if not set,
  then the current context specified in kubeconfig will be used.  It supports
  TLS, username and password, or token-based authentication.  This option is
  ignored if connecting in-cluster (i.e., the endpoint is not specified).
* `namespace` **NAMESPACE** only reads `DNSZone` resources from the namespace
  listed. If this option is omitted then it will read from all namespaces.

>   **Note**: It is recommended to limit the namespace (e.g to `kube-system`)

Reviewer: Ideally we should default to the recommended configuration.  e.g. Default to only reading CRDs from `kube-system`, then perhaps allowing just `namespace` without any namespaces listed mean "all namespaces".

Author: Makes sense.

  because this can be potentially misused. It is ideal to keep the level of
  write access similar to the `coredns` configmap in the `kube-system`
  namespace.

## Ready

This plugin reports readiness to the ready plugin. This will happen after it has
synced to the Kubernetes API.

> 

Reviewer: Should add a section (here or elsewhere) describing how this works with and differs from the Corefile serverblock defined stub domains, and a forward plugin in the same block that covers the following points...
* when DNSZone CRDs and serverblock defined stub domains are used, domains defined in the Corefile take precedence (in the event of zone overlap). e.g. if the domain `example.com` is defined in the corefile as a stub domain, and a DNSZone CRD record defined for `sub.example.com`, then `sub.example.com` would get forwarded to the upstream defined in the corefile, not the DNSZone.
* metrics difference: DNSZone metrics all labeled in single zone (the zone of the enclosing serverblock).
* When using forwardcrd and forward in the same serverblock, DNSZones take precedence over forward plugin defined in the same serverblock. e.g. if a DNSZone is defined for `.`, then no queries would be forwarded to the upstream defined the forward plugin of the same serverblock.

precedence is serverblock -> forwardcrd -> forward

Author: Makes sense. Added to the readme.

  names for remote k8s connection.  This option is ignored if connecting
  in-cluster (i.e. endpoint is not specified).
* `kubeconfig` **KUBECONFIG [CONTEXT]** authenticates the connection to a remote
  k8s cluster using a kubeconfig file.  **[CONTEXT]** is optional, if not set,
  then the current context specified in kubeconfig will be used.  It supports
  TLS, username and password, or token-based authentication.  This option is
  ignored if connecting in-cluster (i.e., the endpoint is not specified).
* `namespace` **[NAMESPACE]** only reads `DNSZone` resources from the namespace
  listed. If this option is omitted then it will read from the default
  namespace, `kube-system`. If this option is specified without any namespaces
  listed it will read from all namespaces.  **Note**: It is recommended to limit
  the namespace (e.g to `kube-system`) because this can be potentially misused.
  It is ideal to keep the level of write access similar to the `coredns`
  configmap in the `kube-system` namespace.

## Ready

This plugin reports readiness to the ready plugin. This will happen after it has
synced to the Kubernetes API.

## Ordering

> Coredns has following precedence:

Reviewer: ```suggestion
Forward behavior can be defined in three ways, via a Server Block, via the *forwardcrd* plugin, and via the *forward* plugin.  If more than one of these methods is employed and a query falls within the zone of more than one, CoreDNS selects which one to use based on the following precedence:
```

> Corefile Server Block -> `forwardcrd` plugin -> `forward` plugin.

Reviewer: ```suggestion
Corefile Server Block -> *forwardcrd* plugin -> *forward* plugin.
```

Reviewer: To be consistent with other plugin documentation, *emphasize* plugin names.

When `DNSZone` CRDs and Server Blocks define stub domains that are used,
domains defined in the Corefile take precedence (in the event of zone overlap).

> in the Corefile take precedence (in the event of zone overlap). e.g. if the

Reviewer: ```suggestion
e.g. if the
```

  listed it will read from all namespaces.  **Note**: It is recommended to limit
  the namespace (e.g to `kube-system`) because this can be potentially misused.
  It is ideal to keep the level of write access similar to the `coredns`
  configmap in the `kube-system` namespace.

## Ready

This plugin reports readiness to the ready plugin. This will happen after it has
synced to the Kubernetes API.

## Ordering

Forward behavior can be defined in three ways, via a Server Block, via the *forwardcrd* plugin, and via the *forward* plugin.  If more than one of these methods is employed and a query falls within the zone of more than one, CoreDNS selects which one to use based on the following precedence:
Corefile Server Block -> *forwardcrd* plugin -> *forward* plugin.

When `Forward` CRDs and Server Blocks define stub domains that are used,
domains defined in the Corefile take precedence (in the event of zone overlap).
e.g. if the
domain `example.com` is defined in the Corefile as a stub domain, and a
`Forward` CRD record defined for `sub.example.com`, then `sub.example.com` would
get forwarded to the upstream defined in the Corefile, not the `Forward` CRD.

> When using `forwardcrd` and `forward` in the same Server Block, `Forward` CRDs
> take precedence over the `forward` plugin defined in the same Server Block.
> e.g. if a `Forward` resource is defined for `.`, then no queries would be
> forwarded to the upstream defined the `forward` plugin of the same Server Block.

Reviewer: ```suggestion
When using *forwardcrd* and *forward* plugins in the same Server Block, `Forward` CRDs
take precedence over the *forward* plugin defined in the same Server Block.
e.g. if a `Forward` CRD is defined for `.`, then no queries would be
forwarded to the upstream defined in the *forward* plugin of the same Server Block.
```

## Metrics

`Forward` CRD metrics are all labeled in a single zone (the zone of the enclosing
Server Block).

> ## Examples

Reviewer: A real-world example of how forwardcrd would be used in an actual CoreDNS Corefile as deployed in a kubernetes cluster would be helpful.

Allow `Forward` resources to be created for any zone and only read `Forward`
resources from the `kube-system` namespace:

~~~ txt
. {
    forwardcrd
}
~~~

Allow `Forward` resources to be created for the `.local` zone and only read
`Forward` resources from the `kube-system` namespace:

~~~ txt
. {
    forwardcrd local
}
~~~

or:

~~~ txt
local {
    forwardcrd
}
~~~

Only read `Forward` resources from the `dns-system` namespace:

~~~ txt
. {
    forwardcrd {
        namespace dns-system
    }
}
~~~

Read `Forward` resources from all namespaces:

~~~ txt
. {
    forwardcrd {
        namespace
    }
}
~~~

Connect to Kubernetes with CoreDNS running outside the cluster:

~~~ txt
. {
    forwardcrd {
        endpoint https://k8s-endpoint:8443
        tls cert key cacert
    }
}
~~~

or:

~~~ txt
. {
    forwardcrd {
        kubeconfig ./kubeconfig
    }
}
~~~

## Forward resource

Apply the `Forward` CRD to your Kubernetes cluster.

```
kubectl apply -f ./manifests/crds/coredns.io_forwards.yaml
```

> Assuming the **forwardcrd** plugin has been configured to allow `Forward`

Reviewer: ```suggestion
Assuming the *forwardcrd* plugin has been configured to allow `Forward`
```
